### ISSUE TITLE
PAL: seh-unwind.cpp causes build to fail on FreeBSD

### ISSUE THREAD

**Issue by josteink** [dotnet/coreclr#639]
seh-unwind.cpp uses direct register-access. The result is that the build fails with the following errors on FreeBSD (excerpt):

````
    ~~~~~~~~~~~~~~~~~~~~~~~ ^
/home/josteink/build/cc2/src/pal/src/exception/seh-unwind.cpp:46:35: error: use of undeclared identifier 'REG_RSP'
    unwContext->uc_mcontext.gregs[REG_RSP] = winContext->Rsp;
                                  ^
/home/josteink/build/cc2/src/pal/src/exception/seh-unwind.cpp:47:29: error: no member named 'gregs' in '__mcontext'
    unwContext->uc_mcontext.gregs[REG_RBP] = winContext->Rbp;
    ~~~~~~~~~~~~~~~~~~~~~~~ ^
/home/josteink/build/cc2/src/pal/src/exception/seh-unwind.cpp:47:35: error: use of undeclared identifier 'REG_RBP'
    unwContext->uc_mcontext.gregs[REG_RBP] = winContext->Rbp;
                                  ^
/home/josteink/build/cc2/src/pal/src/exception/seh-unwind.cpp:48:29: error: no member named 'gregs' in '__mcontext'
    unwContext->uc_mcontext.gregs[REG_RBX] = winContext->Rbx;
````

Should it use the abstractions defined in context.cpp instead?

**Comment by kangaroo** [comment]
What type does unw_context expose on FreeBSD?  Is it an mcontext like Linux?  If so we can probably macro this away.

**Comment by josteink** [comment]
`uc_mcontext` is exposed as `__mcontext` from FreeBSD's `x86/context.h`. This seems to be the exact same context as found in [context.cpp's MCREGs](https://github.com/dotnet/coreclr/blob/master/src/pal/src/arch/i386/context.cpp#L114-L133) for FreeBSD.

That is, for FreeBSD `unwContext->uc_mcontext` (but not `unwContext`) matches the type used for MCREG-mapping in `context.cpp`. From what I can tell it also maps 1-1 to [the other MCREGs](https://github.com/dotnet/coreclr/blob/master/src/pal/src/arch/i386/context.cpp#L56-L76) used by context.cpp for GREGs.

````c++
#if HAVE_GREGSET_T
#ifdef BIT64
#define MCREG_Rbx(mc)       ((mc).gregs[REG_RBX])
#define MCREG_Rcx(mc)       ((mc).gregs[REG_RCX])
...
````

For comparison, the offending code from `seh-unwind.cpp` is [here](https://github.com/dotnet/coreclr/blob/master/src/pal/src/exception/seh-unwind.cpp#L45-L52).

````c++
#if defined(_AMD64_)
    unwContext->uc_mcontext.gregs[REG_RIP] = winContext->Rip;
    unwContext->uc_mcontext.gregs[REG_RSP] = winContext->Rsp;
...
````

Note how it only checks for 64-bit, but unlike context.cpp it doesn't check for GREGs or not.

The best would obviously be reusing what we have in `context.cpp` instead of redefining it here. This does indeed looks like code ripe for re-use.

Any good suggestions for how to get this done as cleanly as possible? C++ is not really my strongest skill as a programmer, and I'm open for suggestions :)

**Comment by janvorli** [comment]
@josteink We should definitely reuse the abstraction in `context.cpp`. It looks like moving the macros to the `context.h` would be the way to go, the `context.h` is already being included from the `seh-unwind.cpp`.
When I have recently added the libunwind support to the seh-unwind.cpp, I have not realized we have these macros covering the same context. That's how we have ended up having it "hardcoded" in the `seh-unwind.cpp`.
